src/Modules/Trinity.FFI/Trinity.FFI/ObjectStore.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Trinity.FFI
+{
+    internal class ObjectStore<T>
+    {
+        // 16 + 8 bytes per entry
+        internal struct AssocT
+        {
+            public T value;
+            public int next;
+        }
+
+        protected AssocT[] m_array = null;
+
+        private long m_pad0 = 0;
+        private long m_pad1 = 0;
+        private long m_pad2 = 0;
+        private long m_pad3 = 0;
+        private long m_pad4 = 0;
+        private long m_pad5 = 0;
+        private long m_pad6 = 0;
+
+        protected int m_head = 0;
+        protected int m_len = 0;
+
+        private long m_pad7 = 0;
+        private long m_pad8 = 0;
+        private long m_pad9 = 0;
+        private long m_pad10 = 0;
+        private long m_pad11 = 0;
+        private long m_pad12 = 0;
+        private long m_pad13 = 0;
+
+        protected object m_lock;
+
+        private const int c_default_len = 1 << 20;
+
+        public ObjectStore()
+        {
+            m_lock = new object();
+            _Alloc_Init(c_default_len);
+            
+        }
+        public Action<int> _Alloc_Mutable;
+
+        private void _Alloc_Init(int newlen)
+        {
+            AssocT[] ret = new AssocT[newlen];
+            for (int init_from = 0; init_from < newlen; ++init_from)
+            {
+                ret[init_from].next = init_from + 1;
+                ret[init_from].value = default(T);
+            }
+            m_array = ret;
+            m_len = m_array.Length;
+            _Alloc_Mutable = _Alloc_Normal;
+        }
+        private void _Alloc_Normal(int newlen)
+        {
+            if (newlen < 0) throw new OutOfMemoryException();
+
+            AssocT[] ret = new AssocT[newlen];
+            int init_from = m_array.Length;
+            m_head = init_from;
+            Array.Copy(m_array, ret, init_from);
+
+            for (; init_from < newlen; ++init_from)
+            {
+                ret[init_from].next = init_from + 1;
+                ret[init_from].value = default(T);
+            }
+            m_array = ret;
+            m_len = m_array.Length;
+        }
+
+        private void _Alloc(int newlen)
+        {
+            _Alloc_Mutable(newlen);
+        }
+
+        public void DeAlloc()
```

Comment:
why public DeAlloc? It means that a user could deallocate the references without disposing them.